crates/attestation/src/crypto/crl.rs

```diff
@@ -0,0 +1,431 @@
+            }
+        }
+        None
+    }
+
+    fn get_crl_by_issuer(&self, issuer: &Name) -> Option<&CertificateList> {
+        for pair in &self.crls {
+            if &pair.crl.tbs_cert_list.issuer == issuer {
+                return Some(&pair.crl);
+            }
+        }
+        None
+    }
+}
+
+pub trait PkiPathCRLCheck<'a> {
+    fn check_crl(&self, pairs: &CrlList<'a>) -> Result<()>;
+}
+
+impl<'a> PkiPathCRLCheck<'a> for PkiPath<'a> {
+    fn check_crl(&self, pairs: &CrlList<'a>) -> Result<()> {
+        // We want to ensure that a valid CRL was passed, so we make sure at least one of the CRLs
+        // is valid for this `PkiPath`, otherwise, no valid CRLs were received, and that's not okay.
+        let mut found = false;
+        let mut iter = self.iter().peekable();
+        loop {
+            let cert = iter.next();
+            let next = iter.peek();
+            if cert.is_none() || next.is_none() {
+                break;
+            }
+            let cert = cert.unwrap();
+            let next = next.unwrap();
+
+            let crls = {
+                let mut crls = vec![];
+                if let Some(crl) = pairs.get_crl_by_issuer(&cert.tbs_certificate.subject) {
+                    crls.push(crl);
+                } else {
+                    let urls = cert.tbs_certificate.get_crl_urls()?;
+                    for url in urls {
+                        if let Some(crl) = pairs.get_crl_by_url(&url) {
+                            crls.push(crl);
+                        }
+                    }
+                }
+
+                crls
+            };
+
+            ensure!(!crls.is_empty(), "Certificate had no CRL URLs");
+
+            for crl in crls {
+                // Don't check CRL expiration in unit tests
+                #[cfg(not(debug_assertions))]
+                if let Some(next_update) = crl.tbs_cert_list.next_update {
+                    if next_update.to_system_time() <= std::time::SystemTime::now() {
+                        bail!("CRL expired");
+                    }
+                }
+
+                let raw_bytes = crl.tbs_cert_list.to_vec().unwrap();
+                match cert.tbs_certificate.subject_public_key_info.verify(
+                    &raw_bytes,
+                    cert.signature_algorithm,
+                    crl.signature.raw_bytes(),
+                ) {
+                    Ok(_) => {
+                        if let Some(revocations) = crl.tbs_cert_list.revoked_certificates.as_ref() {
+                            for revoked in revocations {
+                                if revoked.serial_number == next.tbs_certificate.serial_number {
+                                    bail!("revoked!");
+                                }
+                            }
+                        }
+                        found = true;
+                    }
+                    Err(e) => {
+                        bail!("CRL validation error: {e}");
```

Comment:
Maybe instead do a return `Err(e).context("CRL validation error");`? That way, the underlying error isn't lost, and can be debug-printed if the caller wants to.

Reply:
Done